### PR TITLE
crimson/osd: prevent premature OSD activation.

### PR DESCRIPTION
**Pull request by rzarzynski**
In contrast to the classical OSD:

```
int OSD::init()
{
  // ...

  {
    epoch_t bind_epoch = osdmap->get_epoch();
    service.set_epochs(NULL, NULL, &bind_epoch);
  }

  // ...

  // load up pgs (as they previously existed)
  load_pgs();
```

crimson doesn't set the `bind_epoch` when initializing. The net result is going active prematurely which happens because the 3rd condition (`bind_epoch < osdmap->get_up_from(whoami)`) is always true.

```
    if (osdmap->is_up(whoami) &&
        osdmap->get_addrs(whoami) == public_msgr->get_myaddrs() &&
        bind_epoch < osdmap->get_up_from(whoami)) {
      if (state.is_booting()) {
        logger().info("osd.{}: activating...", whoami);
```

Nullifying it translates the "is it activated?" check basically into "is it up?" verification. This is problematic in a situation
like:

1. Primary got new OSDMap but replica has not.
2. Replica restarts, sends `MOSDBoot` and receives the newer map
   from the previous point.
3. Primary sends a message that is unexpected by replica.
4. Monitor publishes a new OSDMap diven by the `MOSDBoot`.

Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by athanatos — APPROVED** [review]

File: src/crimson/osd/osd.cc [hunk 1]
```diff
@@ -1066,18 +1067,21 @@ seastar::future<> OSD::committed_osd_maps(version_t first,
       }
     });
   }).then([m, this] {
-    if (osdmap->is_up(whoami) &&
-        osdmap->get_addrs(whoami) == public_msgr->get_myaddrs() &&
-        bind_epoch < osdmap->get_up_from(whoami)) {
-      if (state.is_booting()) {
+    if (const auto is_up = osdmap->is_up(whoami); is_up) {
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc:1070]:
```suggestion
    if (osdmap->is_up(whoami)) {
```

File: src/crimson/osd/osd.cc [hunk 2]
```diff
@@ -1066,18 +1067,21 @@ seastar::future<> OSD::committed_osd_maps(version_t first,
...
     });
   }).then([m, this] {
-    if (osdmap->is_up(whoami) &&
-        osdmap->get_addrs(whoami) == public_msgr->get_myaddrs() &&
-        bind_epoch < osdmap->get_up_from(whoami)) {
-      if (state.is_booting()) {
+    if (osdmap->is_up(whoami)) {
+      const auto up_from = osdmap->get_up_from(whoami);
+      logger().debug("osd.{} i'm up! up_from {}, bind_epoch {}, state {}",
+                     whoami, up_from, bind_epoch, state);
```
Comment (tchaikov) [inline src/crimson/osd/osd.cc:1073]:
```c++
logger().info("osd.{}: map e {} marked me up: up_from {}, bind_epoch {}, state {}",
              whoami, osdmap->get_epoch(), up_from, bind_epoch, state));
```

**Review by tchaikov — APPROVED** [review]

File: src/crimson/osd/osd.cc (hunk 2 above)
Comment (tchaikov) [reply src/crimson/osd/osd.cc:1073]: @rzarzynski just a nit, but i don't feel strong either way. please feel free to merge this PR.

Comment (rzarzynski) [reply src/crimson/osd/osd.cc]: Amended!

**Pull request merged**